src/goto-instrument/aggressive_slicer.h

```diff
@@ -0,0 +1,100 @@
+/*******************************************************************\
+
+Module: Aggressive Slicer
+
+Author: Elizabeth Polgreen, polgreen@amazon.com
+
+\*******************************************************************/
+
+/// \file
+/// Aggressive slicer
+/// Consider the control flow graph of the goto program and a criterion
+/// description of aggressive slicer here
+
+#ifndef CPROVER_GOTO_INSTRUMENT_AGGRESSIVE_SLICER_H
+#define CPROVER_GOTO_INSTRUMENT_AGGRESSIVE_SLICER_H
+
+#include <list>
+#include <string>
+#include <unordered_set>
+
+#include <util/irep.h>
+
+#include <analyses/call_graph.h>
+
+#include <linking/static_lifetime_init.h>
+
+class goto_modelt;
+class message_handlert;
+
+/// \brief The aggressive slicer removes the body of all the functions except
+/// those on the shortest path, those within the call-depth of the
+/// shortest path, those given by name as command line argument,
+/// and those that contain a given irep_idt snippet
+/// If no properties are set by the user, we preserve all functions on
+/// the shortest paths to each property.
+class aggressive_slicert
+{
+public:
+  aggressive_slicert(goto_modelt &_goto_model, message_handlert &_msg)
+    : preserve_all_direct_paths(false),
+      goto_model(_goto_model),
+      message_handler(_msg)
+  {
+    start_function = goto_model.goto_functions.entry_point();
+    call_depth = 0;
+    call_grapht undirected_call_graph =
+      call_grapht::create_from_root_function(goto_model, start_function, false);
+    call_graph = undirected_call_graph.get_directed_graph();
+  }
+
+  ///  \brief Removes the body of all functions except those on the
+  /// shortest path or functions
+  /// that are reachable from functions on the shortest path within
+  /// N calls, where N is given by the parameter "distance"
+  void doit();
+
+  /// \brief Adds a list of functions to the set of functions for the aggressive
+  /// slicer to preserve
+  /// \param  function_list:  a list of functions in form
+  /// std::list<std::string>, as returned by get_cmdline_option.
+  void preserve_functions(const std::list<std::string> &function_list)
+  {
+    for(const auto &f : function_list)
+      functions_to_keep.insert(f);
+  };
+
+  std::list<std::string> user_specified_properties;
```

Comment:
Why are these public?

Reply:
these are written to in goto_instrument_parse_options.cpp